### PR TITLE
[Master] Fix incorrect cyclic function detection

### LINKED ISSUE #40578: [Bug]: Invalid cyclic ref error for module level function
### Description

When there is a function call for another module and if there is a function definition in same name, the compiler detects a cyclic ref error.

### Steps to Reproduce

```ballerina
import ballerina/edi;

type EdiDeserialize function ();

public function fromEdi835String() {
    json|error res = edi:fromEdiString("ediText", {delimiters: {segment: "", 'field: "", component: ""}, name: ""});
}

final readonly & map<EdiDeserialize> ediDeserializers = {
    "835": fromEdi835String
};

public isolated function fromEdiString() { // function same to `fromEdiString`
    EdiDeserialize? ediDeserialize = ediDeserializers["ediName"];
    if ediDeserialize is () {
        return ();
    }
    _ = ediDeserialize();
}
```

### Affected Version(s)

U5

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose

Fixes #40578

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/functions/test_different_function_signatures.bal [hunk 1]
```diff
@@ -326,6 +326,24 @@ function testMethodVarargEvaluationCount() {
...
 }
 
+type Func function ();
+
+public function f1() {
+    json|error res = foo:testFunc("f1");
+}
+
+final readonly & map<Func> func = {
+    "f1": f1
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/functions/test_different_function_signatures.bal:336]: I think this should ideally work even if `f1` calls `testFunc` defined in this module - https://github.com/ballerina-platform/ballerina-lang/issues/30042.

Shall we add another test where it is calling the function?

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+type Func isolated function ();
+
+final readonly & map<Func> func = {
+    "f1": f1
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal]: Changed

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_project/different_function_signatures.bal [hunk 3]
```diff
@@ -143,3 +143,7 @@ public client class FooTwo {
         return [i, s, checkpanic t.cloneWithType(StringArray)];
     }
 }
+
+public isolated function testFunc(string ediText) returns json|error {
+    return {};
+}
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/bala/test_projects/test_project/different_function_signatures.bal]: Changed

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/functions/test_different_function_signatures.bal [hunk 4]
```diff
@@ -326,6 +326,36 @@ function testMethodVarargEvaluationCount() {
...
+    if res is error {
+        return -1;
+    }
+    return 1;
+}
+
+final readonly & map<int> func1 = {
+    "f1": f2()
+};
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/functions/test_different_function_signatures.bal:358]: Why aren't we asserting anything?

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+final readonly & map<json|error> func = {
+    "f1": f1()
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal:18]: I would add the test without the call too. To ensure the current behaviour is tested. So two tests, one which may no longer result in an error in the future.

Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal:18]: https://github.com/ballerina-platform/ballerina-lang/blob/3d985ba1a71292f872ed5a1a1e21bf672fce047a/tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/functions/test_different_function_signatures.bal#L336
Here we have a test like that.

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/functions/test_different_function_signatures.bal (hunk 4 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/functions/test_different_function_signatures.bal:358]: Added

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal (hunk 5 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal:18]: Not a negative test though?

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+};
+
+public isolated function testFunc() {
+    json|error? f = func["f1"];
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal:22]: Do we need nil in the type?

Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal]: Changed

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal (hunk 5 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/functions/cyclic-functions-negative.bal:18]: Added a test

**Review by gimantha — APPROVED** [review]

**Pull request merged**